src/ucunit/include/ucunit/ucunit.h

```diff
@@ -30,225 +30,140 @@ extern "C" {
 #define CU_UNREACHABLE ((void)0)
 #endif
 
-#define CU_ASSERT_IMPLEMENTATION(value_, line, expr, file, something, fatal_) do { \
-  const bool cu_assert_impl_value = (value_); \
-  const bool cu_assert_impl_fatal = (fatal_); \
-  CU_assertImplementation (cu_assert_impl_value, (line), (expr), (file), (something), cu_assert_impl_fatal); \
-  if (!cu_assert_impl_value && cu_assert_impl_fatal) \
+#define CU_PASS(msg) \
+  CU_assertImplementation (true, __LINE__, ("CU_PASS(" msg ")"), __FILE__, "", false)
+#define CU_FAIL(msg) \
+  CU_assertImplementation (false, __LINE__, ("CU_FAIL(" msg ")"), __FILE__, "", false)
+#define CU_FAIL_FATAL(msg) do { \
+    CU_assertImplementation (false, __LINE__, ("CU_FAIL_FATAL(" msg ")"), __FILE__, "", true); \
     CU_UNREACHABLE; \
+  } while (0)
+
+#if defined __STDC__ && __STDC_VERSION__ >= 201100L
+
+#define CU_ASSERT_PRINTF_FORMAT(T_)             \
+  _Generic ((T_),                               \
+    _Bool             : "%d",                   \
+    char              : "%c",                   \
+    signed char       : "%hhd",                 \
+    unsigned char     : "%hhu",                 \
+    short             : "%hd",                  \
+    int               : "%d",                   \
+    long              : "%ld",                  \
+    long long         : "%lld",                 \
+    unsigned short    : "%hu",                  \
+    unsigned int      : "%u",                   \
+    unsigned long     : "%lu",                  \
+    unsigned long long: "%llu",                 \
+    float             : "%f",                   \
+    double            : "%f",                   \
+    long double       : "%Lf",                  \
+    default           : "%p"                    \
+  )
+
+#define CU_ASSERT_OP_MAYBE_FATAL(x_, op_, y_, fatal_) do {              \
+  typeof (x_) xv__ = (x_);                                              \
+  typeof (x_) yv__ = (y_); /* note: typeof x_ here, not typeof y_) */   \
```

Comment:
Casting `y_` to the type of `x_` seems a bit tricky, e.g. in
```
    uint8_t c = 'c';
    uint32_t x = 256 + (uint32_t) c;
    CU_ASSERT_EQ_FATAL(c, x);
```
the compiler will complain with a warning on the cast (that of course shouldn't be ignored), but the test succeeds.